src/sbf.cpp

```diff
@@ -1,590 +1,713 @@
+
+		sendMessageAndWaitForAck(msg, SBF_CONFIG_TIMEOUT);
+	}
+
+	decodeInit();
+	receive(50);
+	decodeInit();
+
+	// Output a set of SBF blocks on a given connection at a regular interval.
+	int i = 0;
+	snprintf(msg, sizeof(msg), SBF_CONFIG, com_port);
+
+	do {
+		++i;
+
+		if (!sendMessageAndWaitForAck(msg, SBF_CONFIG_TIMEOUT)) {
+			if (i >= 5) {
+				return -1; // connection and/or baudrate detection failed
+			}
+
+		} else {
+			response_detected = true;
+		}
+	} while (i < 5 && !response_detected);
+
+
+
+	if (_output_mode == OutputMode::RTCM) {
+		if (_base_settings.type == BaseSettingsType::fixed_position) {
+			snprintf(msg, sizeof(msg), SBF_CONFIG_RTCM_STATIC_COORDINATES,
+				 _base_settings.settings.fixed_position.latitude,
+				 _base_settings.settings.fixed_position.longitude,
+				 static_cast<double>(_base_settings.settings.fixed_position.altitude));
+			sendMessageAndWaitForAck(msg, SBF_CONFIG_TIMEOUT);
+			snprintf(msg, sizeof(msg), SBF_CONFIG_RTCM_STATIC_OFFSET, 0.0, 0.0, 0.0);
+			sendMessageAndWaitForAck(msg, SBF_CONFIG_TIMEOUT);
+			sendMessageAndWaitForAck(SBF_CONFIG_RTCM_STATIC1, SBF_CONFIG_TIMEOUT);
+			sendMessageAndWaitForAck(SBF_CONFIG_RTCM_STATIC2, SBF_CONFIG_TIMEOUT);
+		}
+	}
+
+	_configured = true;
+	return 0;
+}
+
+bool GPSDriverSBF::sendMessage(const char *msg)
+{
+	// Send message
+	SBF_DEBUG("Send MSG: %s", msg);
+	int length = static_cast<int>(strlen(msg));
+
+	return (write(msg, length) == length);
+}
+
+bool GPSDriverSBF::sendMessageAndWaitForAck(const char *msg, const int timeout)
+{
+	SBF_DEBUG("Send MSG: %s", msg);
+
+	// Send message
+	int length = static_cast<int>(strlen(msg));
+
+	if (write(msg, length) != length) {
+		return false;
+	}
+
+	// Wait for acknowledge
+	// For all valid set -, get - and exe -commands, the first line of the reply is an exact copy
+	// of the command as entered by the user, preceded with "$R:"
+	char buf[GPS_READ_BUFFER_SIZE];
+	size_t offset = 1;
+	gps_abstime time_started = gps_absolute_time();
+
+	bool found_response = false;
+
+	do {
+		--offset; //overwrite the null-char
+		int ret = read(reinterpret_cast<uint8_t *>(buf) + offset, sizeof(buf) - offset - 1, timeout);
+
+		if (ret < 0) {
+			// something went wrong when polling or reading
+			SBF_WARN("sbf poll_or_read err");
+			return false;
+		}
+
+		offset += ret;
+		buf[offset++] = '\0';
+
+		if (!found_response && strstr(buf, "$R: ") != nullptr) {
+			//SBF_DEBUG("READ %d: %s", (int) offset, buf);
+			found_response = true;
+		}
+
+		if (offset >= sizeof(buf)) {
+			offset = 1;
+		}
+
+	} while (time_started + 1000 * timeout > gps_absolute_time());
+
+	SBF_DEBUG("response: %u", found_response)
+	return found_response;
+}
+
+// -1 = error, 0 = no message handled, 1 = message handled, 2 = sat info message handled
+int GPSDriverSBF::receive(unsigned timeout)
+{
+	// Do not receive messages until we're configured
+	if (!_configured) {
+		gps_usleep(timeout * 1000);
+		return 0;
+	}
+
+	uint8_t buf[RECEIVE_BUF_SIZE];

```

Comment:
Why was this changed from GPS_READ_BUFFER_SIZE?

Reply:
The GPS_READ_BUFFER_SIZE is too big which causes the board to crash. I don't know if there is a fix for this, but the smaller size is enough and the program still works.

Reply:
`GPS_READ_BUFFER_SIZE` is smaller:
```
#define RECEIVE_BUF_SIZE 800
#define GPS_READ_BUFFER_SIZE 150
```

Reply:
Yes, the GPS_READ_BUFFER_SIZE is large enough for the cmd replies but not enough for actual position data

Reply:
It does not have to be, it's assembled individually via `parseChar`

Reply:
True, i changed it to GPS_READ_BUFFER_SIZE

Reply:
True, but if the buffer is not large enough, it will not always recognize an SBF block so the timeout will cause the receiver to be reinitialized.

Reply:
Can you debug this, it does not make sense.
This looks incorrect: https://github.com/PX4/PX4-GPSDrivers/blob/ebaed6a08b8bdcf278144b4d79815115bbc8fae2/src/sbf.cpp#L463= (should be 0)
And the return value from payloadRxAdd should not be passed back to parseChar: https://github.com/PX4/PX4-GPSDrivers/blob/ebaed6a08b8bdcf278144b4d79815115bbc8fae2/src/sbf.cpp#L358=
https://github.com/PX4/PX4-GPSDrivers/blob/ebaed6a08b8bdcf278144b4d79815115bbc8fae2/src/sbf.cpp#L373=
You can compare with the ubx driver.

Reply:
Never mind, it does work with GPS_READ_BUFFER_SIZE. I had to many output-streams on the receiver running